scout/server/extensions/beacon_extension.py

```diff
@@ -0,0 +1,192 @@
+
+
+class Beacon:
+    """Interface to cgbeacon2 server instance, reachable via REST API"""
+
+    def __init__(self):
+        self.add_variants_url = None
+        self.delete_variants_url = None
+        self.token = None
+
+    def init_app(self, app):
+        """Initialize the beacon extension and make its parametars available to the app."""
+        self.token = app.config.get("BEACON_TOKEN")
+        self.beacon_url = app.config.get("BEACON_URL")
+
+        self.add_variants_url = "/".join([self.beacon_url, "add"])
+        self.delete_variants_url = "/".join([self.beacon_url, "delete"])
+
+    def get_datasets(self):
+        """Makes a call to the Beacon's info endpoint (/) and extracts a complete list of available datasets
+        Returns:
+            datasets(list): list of dataset _ids. Example: ["cust002_GRCh37", "cust000_GRCh38", ..]
+        """
+        datasets = []
+        json_resp = get_request_json(url=f"{self.beacon_url}/")
+        if json_resp.get("status_code") == 200:
+            datasets = json_resp.get("content", {}).get("datasets", [])
+        else:
+            flash("Error retrieving Beacon's dataset list:{json_resp}")
+        return [dset["id"] for dset in datasets]
+
+    def base_submission_data(self, store, case_obj, form):
+        """Create data dictionary to be send as json data in a POST request to the beacon "add" endpoint
+
+        Args:
+            store(adapter.MongoAdapter)
+            case_obj(dict): scout.models.Case
+            form(ImmutableMultiDict): request form submitted by user. Example:
+                [('case', 'internal_id'), ('samples', 'affected'), ('vcf_files', 'vcf_snv'), ('vcf_files', 'vcf_snv_research'), ('panels', '6246b25121d86882e127710c')]
+
+        Returns:
+            data(dict): a dictionary with base info to be use as json data in beacon add request (lacks path to VCF file to extract variants from)
+        """
+        # Initialize key/values to be sent in request:
+        assembly = "GRCh38" if "38" in str(case_obj.get("genome_build", "37")) else "GRCh37"
+        dataset_id = "_".join([case_obj["owner"], case_obj.get("build", assembly)])
+
+        samples = []
+        if form.get("samples") == "affected":
+            samples = [
+                ind["individual_id"]
+                for ind in case_obj.get("individuals", [])
+                if ind["phenotype"] == 2
+            ]
+        else:
+            samples = [ind["individual_id"] for ind in case_obj.get("individuals")]
+
+        data = {
+            "dataset_id": dataset_id,
+            "samples": samples,
+            "assemblyId": assembly,
+        }
+
+        gene_ids = set()
+        for panel in form.getlist("panels"):
+            gene_ids.update(store.panel_to_genes(panel_id=panel, gene_format="hgnc_id"))
+
+        if gene_ids:
+            data["genes"] = {"ids": list(gene_ids), "id_type": "HGNC"}
+
+        return data
+
+    def add_variants(self, store, case_obj, form):
```

Comment:
Passing the form all the way to the extension does connect things quite a bit. But the code is currently clean enough that I think changing it to parse the form in the view would make it more complex. All good for now!

Reply:
Did you mean that it is fine to pass the raw request of it should be parsed in the view first? Sorry i didn't get it

Reply:
I think it is in principle more correct if the view parses requests - or possibly a dedicated function from the controller. But we don’t hold to that in general, and this code is still very clean - the forms are not touched until they reach the extension, so I do not mind an exception!

Reply:
Ah OK, thanks!